repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+        val contactId = (json \ "access" \ "user" \ "RAX-AUTH:contactId").asOpt[String]
+        val expirationDate = iso8601ToRfc1123((json \ "access" \ "token" \ "expires").as[String])
+        val impersonatorId = (json \ "access" \ "RAX-AUTH:impersonator" \ "id").asOpt[String]
+        val impersonatorName = (json \ "access" \ "RAX-AUTH:impersonator" \ "name").asOpt[String]
+        val validToken = ValidToken(expirationDate, userId, username, tenantName, defaultTenantId, tenantIds, roleNames, impersonatorId, impersonatorName, defaultRegion, contactId)
+
+        Option(config.getCache) foreach { cacheSettings =>
+          // never null because configurationUpdated in KeystoneV2Filter
+          val timeout = offsetTtl(cacheSettings.getTimeouts.getToken, cacheSettings.getTimeouts.getVariability)
+          datastore.put(s"$TOKEN_KEY_PREFIX$token", validToken, timeout, TimeUnit.SECONDS)
+        }
+        Success(validToken)
+      } catch {
+        case oops@(_: JsResultException | _: JsonProcessingException) =>
+          Failure(IdentityCommuncationException("Unable to parse JSON from identity validate token response", oops))
+      }
+    }
+
+    val identityEndpoint = config.getIdentityService.getUri
+
+    import scala.collection.JavaConverters._
+    Try(akkaServiceClient.get(s"$TOKEN_KEY_PREFIX$token",
+      s"$identityEndpoint$TOKEN_ENDPOINT/$token",
+      (Map(CommonHttpHeader.AUTH_TOKEN.toString -> authenticatingToken,
+        CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava)
+    ) match {
+      case Success(serviceClientResponse) =>
+        //DEAL WITH IT
+        //Parse the response for validating a token?
+        logger.debug(s"SERVICE CLIENT RESPONSE: ${serviceClientResponse.getStatus}")
+        logger.debug(s"Admin Token: $authenticatingToken")
+        serviceClientResponse.getStatus match {
+          case SC_OK | SC_NON_AUTHORITATIVE_INFORMATION =>
+            //Extract the roles from the JSON and stick it in the ValidToken result
+            extractUserInformation(serviceClientResponse.getData)
+          case SC_BAD_REQUEST => Failure(IdentityValidationException("Bad Token Validation request to identity!"))
+          case SC_UNAUTHORIZED => Failure(AdminTokenUnauthorizedException("Unable to validate token, authenticating token unauthorized"))
+          case SC_FORBIDDEN => Failure(IdentityAdminTokenException("Admin token unauthorized to validate token"))
+          case SC_NOT_FOUND =>
+            Option(config.getCache) foreach { cacheSettings =>
+              // never null because configurationUpdated in KeystoneV2Filter
+              val timeout = offsetTtl(cacheSettings.getTimeouts.getToken, cacheSettings.getTimeouts.getVariability)
+              datastore.put(s"$TOKEN_KEY_PREFIX$token", InvalidToken, timeout, TimeUnit.SECONDS)
```

Comment:
Don't do this. this may become a legit token and you've just locked it out till it timeouts